Implement advanced request mapping

dongjinleekr opened the pull request:
This pull request resolves #150

dongjinleekr commented:
Hello. This is my work which resolves #150. Although there are lots of things to be improved before merging, I thought it is much better to share my (dirty) draft as soon as possible, for following reasons:

1. This issue is much more complex than expected, to the extent that it generates several sub-issues. (detailed descriptions are below.)
2. By above reason, conducting interim check would be better.

My approach is simple: From a given (Spring-MVC like) annotated strategy class, HttpDynamicService generates a list of entries which can capture the path variables and invoke the corresponding method with reflection. (HttpDynamicService#forStrategy). for every request, HttpDynamicService looks for the bindable path from the entries (DynamicPath#matches), deserialize the parameters to adequate types (DynamicPath#bind, DynamicMethod#arguments), and returns the object from DynamicMethod#invoke.

This work generates following sub-issues:

1. Extending RequestMapping to handle the HTTP methods. (see: [RequestMapping annotation in Spring MVC](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestMapping.html))
2. Improved deserializing methods should be supported: Currently, only primitive + String type parameters can be bound. It is much better to support custom deserializing strategy, which can generate any class object from the path variables. (Add to this, current deserialization routine is not only dirty but also bad.)
3. Improved serializing methods should be supported: Currently, all responses are returned in text/plain form.

I am looking forward to your feedback. Thank you very much.

trustin commented:
Wow, looks very promising. Let me take a closer look a little bit later.

trustin commented on core/src/main/java/com/linecorp/armeria/server/ParameterPathMapping.java:
```diff
@@ -0,0 +1,28 @@
+package com.linecorp.armeria.server;
+
+final class ParameterPathMapping extends AbstractPathMapping {
```
This seems unused?

trustin commented on core/src/main/java/com/linecorp/armeria/annotations/PathVariable.java:
```diff
@@ -0,0 +1,18 @@
+package com.linecorp.armeria.annotations;
```
Given that these annotations are only applicable to the strategies given to `HttpDynamicService`, how about moving them to the same package with `HttpDynamicService`?

trustin commented on core/src/main/java/com/linecorp/armeria/annotations/RequestMapping.java:
```diff
@@ -0,0 +1,18 @@
+package com.linecorp.armeria.annotations;
```
Global comment: missing copyright headers

trustin commented on core/src/main/java/com/linecorp/armeria/annotations/package-info.java:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2015 LINE Corporation
```
Global comment: It's not 2015 :-)

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @see DynamicMethod
+ */
+final class Deserializers {
```
Question: We eventually want to write something much more flexible than this, right? HTTP content type negotiation and content conversion, etc.

dongjinleekr replied:
Of course. I will show you more updated, flexible one for this feature next.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/HttpDynamicService.java:
```diff
@@ -0,0 +1,118 @@
+/**
+ * An {@link HttpService} that serves dynamic contents.
+ */
+public final class HttpDynamicService extends AbstractHttpService {
```
`DynamicHttpService`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/HttpDynamicService.java:
```diff
@@ -0,0 +1,118 @@
+            res.respond(HttpStatus.NOT_FOUND);
+        } else {
+            try {
+                Object returned = invokable.get().invoke(path);
```
Given that this is a blocking method, you need to run it in `ctx.blockingTaskExecutor()`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicMethod.java:
```diff
@@ -0,0 +1,101 @@
+     * parameters of the method is not annotated with {@link PathVariable} or deserializable type, raises
+     * {@link IllegalArgumentException}.
+     * */
+    public static DynamicMethod of(Object object, Method method) {
```
Could we also add support for asynchronous methods? For example, we could detect a method that returns a `CompletionStage` or a `CompletableFuture`, and run it without leaving the I/O thread if so.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/HttpDynamicService.java:
```diff
@@ -0,0 +1,118 @@
+
+                HttpHeaders headers = HttpHeaders.of(HttpStatus.OK)
+                                                 .set(HttpHeaderNames.CONTENT_TYPE,
+                                                      MediaType.PLAIN_TEXT_UTF_8.type())
```
I'm not sure if it's OK to send this header given that we do not know the actual content type. Let's remove it.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/HttpDynamicService.java:
```diff
@@ -0,0 +1,118 @@
+                HttpHeaders headers = HttpHeaders.of(HttpStatus.OK)
+                                                 .set(HttpHeaderNames.CONTENT_TYPE,
+                                                      MediaType.PLAIN_TEXT_UTF_8.type())
+                                                 .set(HttpHeaderNames.CONTENT_ENCODING, Charsets.UTF_8.name())
```
`content-encoding` header is not where you set the character set of the content. Let's remove it.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/HttpDynamicService.java:
```diff
@@ -0,0 +1,118 @@
+                                                 .set(HttpHeaderNames.CONTENT_ENCODING, Charsets.UTF_8.name())
+                                                 .setInt(HttpHeaderNames.CONTENT_LENGTH, data.length())
+                                                 .setTimeMillis(HttpHeaderNames.DATE, current)
+                                                 .setTimeMillis(HttpHeaderNames.LAST_MODIFIED, current);
```
We don't really know when the content was modified last. Let's remove it.

dongjinleekr commented:
I totally agree with you. However, I have a question: May I add a new method in ServiceRequestContext to store bound path variables? This method is needed in HttpService#serve method, to retrieve bound variables.

trustin commented:
> May I add a new method in ServiceRequestContext to store bound path variables? This method is needed in HttpService#serve method, to retrieve bound variables.

You can use an attribute. `ServiceRequestContext` implements `AttributeMap`, and thus you can attach an arbitrary attribute to it.

dongjinleekr commented:
Hello @trustin, I updated my PR with your comments. The main difference with the previous work is moving to builder-pattern-based service construction, referring [SparkJava](http://sparkjava.com/).

Following is the list of additional todo-tasks:
1. Enable configuring HTTP Method in @RequestMapping annotation.
2. Implement improved Content-Type feature.
3. Enable asynchronous method invocation: In my opinion, this feature is in fact so critical, since lots of Tomcat based (legacy) microservices are suffering from lack of asynchronous method invocation. If we can support this feature so successfully, it can help their migration to Armeria.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,165 @@
+ * A bindable path. It holds three things:
+ *
+ * <p>
+ * 1. A regex-compiled form of the path. It is used for matching and binding.
```
Use `<ul>` and `<li>` tags

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceConfig.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    void validate(DynamicHttpFunctionEntry entry) {
+        requireNonNull(entry, "DynamicMethodEntry");
```
`"entry"`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * Set of the variables declared in this dynamic path.
+     * */
+    private final Set<String> variables = Sets.newHashSet();
```
Build an immutable set.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,165 @@
+    /**
+     * Returns Binding results with given path.
+     * */
+    public Map<String, String> bind(String path) {
```
Instead of having `matches()` and `bind()`, how about merging these into a single method for efficiency? If not matches, the method could return `null`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Provides various utility functions for {@link Method} object.
+ */
+public final class Methods {
```
Could be package-local?

trustin commented on core/src/test/java/com/linecorp/armeria/server/http/HttpServiceTest.java:
```diff
@@ -91,12 +104,41 @@ protected void doGet(
                       }.decorate(LoggingService::new));
+            sb.service(PathMapping.ofPrefix("/dynamic"), new DynamicHttpServiceBuilder()
+                    .add(HttpMethod.POST, "/{var}", (ctx, req) -> {
+                             final DefaultHttpResponse res = new DefaultHttpResponse();
```
Instead of using `HttpService`, how about defining a new interface that's more suitable for dynamic HTTP services? For example:

```java
@FunctionalInterface
public interface .... {
    void serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> params, HttpResponseWriter res);
}
```

dongjinleekr commented:
I applied all your comments and added HTTP Method support, following Spring MVC's [RequestMapping annotation](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestMapping.html). DynamicHttpFunction is a dedicated functional interface for dynamic HTTP services now. And I removed DynamicHttpServerConfig, for it only contains List of DynamicHttpFunctionEntry - it is useless.

The followings are remaining issues:

- Support Content-Type specification feature in @RequestMapping[^1].
- Make use of Java 8's callback functionality in order to separate main thread from the execution of DynamicHttpFunction[^2].

[^1]: In the case of Spring MVC's [RequestMapping annotation](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestMapping.html), 'produces' attribute takes care of this work.
[^2]: For example, if the DynamicHttpFunction tries to access the other web resources served by poor machine, current blocking-implementation can be a serious problem.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunction.java:
```diff
@@ -0,0 +1,18 @@
+ */
+@FunctionalInterface
+public interface DynamicHttpFunction {
+    HttpResponse serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> bound)
```
- bound -> params or args
- How about adding `HttpResponseWriter` parameter to this method so that a user does not have to create a new `DefaultHttpResponse` by him/herself.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,83 @@
+    public MappedDynamicFunction bind(HttpMethod method, String mappedPath) {
+        if (!this.methods.contains(method)) {
+            return null;
+        } else {
```
else is redundant.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,83 @@
+            return null;
+        } else {
+            Map<String, String> bound = path.bind(mappedPath);
+            if (null != bound) {
```
How about returning null first?

```java
if (bound == null) {
    return null;
}

return new MappedDynamicFunction(bound, function);
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,83 @@
+        if (!this.methods.contains(method)) {
+            return null;
+        } else {
+            Map<String, String> bound = path.bind(mappedPath);
```
bound -> params or args?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,178 @@
+    /**
+     * Parameter entry, which will be used to invoke the method.
+     * */
+    private static class ParameterEntry {
```
Please move to the bottom.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,178 @@
+                                                               .filter(e -> e.overlaps(entry))
+                                                               .findFirst();
+        if (overlapped.isPresent()) {
+            throw new IllegalArgumentException();
```
Please add description.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/MappedDynamicFunction.java:
```diff
@@ -0,0 +1,37 @@
+     * */
+    private final DynamicHttpFunction function;
+
+    MappedDynamicFunction(Map<String, String> bound, DynamicHttpFunction function) {
```
How about making `function` the first parameter?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/PathVariable.java:
```diff
@@ -0,0 +1,34 @@
+
+/**
+ * Class for mapping dynamic web requests onto specific parameter.
+ * */
```
` */`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/package-info.java:
```diff
@@ -0,0 +1,20 @@
+ */
+
+/**
+ * Dynamic http service.
```
HTTP

trustin commented on core/src/test/java/com/linecorp/armeria/server/http/HttpServiceTest.java:
```diff
@@ -91,12 +102,39 @@ protected void doGet(
                           }
                       }.decorate(LoggingService::new));
+            sb.service(PathMapping.ofPrefix("/dynamic"), new DynamicHttpServiceBuilder()
+                    .add(HttpMethod.POST, "/{var}", (ctx, req, arguments) -> {
```
Consistency? args, bound, params - choose one.

trustin commented on core/src/test/java/com/linecorp/armeria/server/http/HttpServiceTest.java:
```diff
@@ -91,12 +102,39 @@ protected void doGet(
     }
 
+    public static class ResponseStrategy {
+        @RequestMapping(method = { HttpMethod.GET, HttpMethod.POST }, value = "/{message}/with/{count}")
```
How about adding dedicated method annotations such as `@GET`, `@POST` instead for brevity? The resulting annotations could be:

```java
@GET @POST
@RequestMapping("/{message}/with/{count}")
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/RequestMapping.java:
```diff
@@ -0,0 +1,44 @@
+ * */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface RequestMapping {
```
How about renaming to `@Path`? (Yes, I like the shortness of JAX-RS annotations :-))

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/PathVariable.java:
```diff
@@ -0,0 +1,34 @@
+ * */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+public @interface PathVariable {
```
How about `@PathParam` which is shorter?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,146 @@
+        StringJoiner skeletonJoiner = new StringJoiner("/");
+        Set<String> placeholders = Sets.newHashSet();
+        for (String token : str.split("/")) {
+            if (token.startsWith("{") && token.endsWith("}")) {
```
Could we also support the pattern of `:var`? e.g. `/:x/:y/:z`? (because it's simpler and as popular as `{...}`)

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpService.java:
```diff
@@ -0,0 +1,62 @@
+
+        Optional<MappedDynamicFunction> mapped = entries.stream().map(e -> e.bind(method, path))
+                                                        .filter(Objects::nonNull)
+                                                        .findFirst();
```
Given that this may be a hot spot, could we avoid using a stream? How about making `entries` member field a `DynamicHttpFunctionEntry[]` so that it's as cheap as possible to loop over it?

dongjinleekr replied:
No problem.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,178 @@
+ * Builds a new {@link DynamicHttpService}.
+ * */
+public class DynamicHttpServiceBuilder {
+    private final List<DynamicHttpFunctionEntry> entries = Lists.newArrayList();
```
`new ArrayList<>();`? No reason to use Guava here.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,178 @@
+    /**
+     * Adds a new dynamic-mapped method, invoked with given {@link HttpMethod} and (dynamically-bound) path.
+     * */
+    public DynamicHttpServiceBuilder add(HttpMethod method, String path, DynamicHttpFunction function) {
```
Could we add the builder methods for all HTTP methods? e.g.

```java
builder.get("/some_path", (ctx, req, params, res) -> { ... });
```

trustin commented:
> Support Content-Type specification feature in @RequestMapping
> Make use of Java 8's callback functionality in order to separate main thread from the execution of DynamicHttpFunction

Both are related with each other in my opinion.

Currently, DynamicHttpFunction is expected to return an `HttpResponse`, but we could make it return anything, and we could behave differently depending on the type of the returned value:

- `CompletionStage<T>` - resolve the returned future asynchronously. When the future is complete, we convert T into an `HttpResponse` using one of the conversion mechanisms available. JAX-RS and Spring Web already have well-written API for content negotiation and conversion, so we could get some idea from what they have achieved.
- `HttpResponse` - pass-through
- Others - invoke `DynamicHttpFunction` in a `RequestContext.blockingTaskExecutor()` unless it's annotated with a special annotation such as `@NonBlocking`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Connect.java:
```diff
@@ -0,0 +1,32 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface Connect {
```
I guess we don't need this because we do not support this method.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * {@link HttpMethod}, {@link DynamicPath} and their corresponding {@link DynamicHttpFunction} instance.
+ */
+class DynamicHttpFunctionEntry {
```
`final`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Returns whether it's mapping overlaps with given {@link DynamicHttpFunctionEntry} instance.
+     */
+    public boolean overlaps(DynamicHttpFunctionEntry entry) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @see MappedDynamicFunction
+     */
+    public MappedDynamicFunction bind(HttpMethod method, String mappedPath) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpService.java:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * An array of {@link DynamicHttpFunctionEntry}.
+     */
+    private DynamicHttpFunctionEntry[] entries;
```
`final`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpService.java:
```diff
@@ -0,0 +1,59 @@
+public final class DynamicHttpService extends AbstractHttpService {
+
+    /**
+     * An array of {@link DynamicHttpFunctionEntry}.
```
Global comment: this comment adds nothing.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+/**
+ * Builds a new {@link DynamicHttpService}.
+ */
+public class DynamicHttpServiceBuilder {
```
Could be `final`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+    /**
+     * Adds a new dynamic-mapped method, invoked with given {@link HttpMethod}s and (dynamically-bound) path.
+     */
+    public DynamicHttpServiceBuilder add(EnumSet<HttpMethod> methods, String path,
```
EnumSet -> Iterable

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+
+            DynamicHttpFunction function = new DynamicHttpFunction() {
+                /**
+                 * Returns parameters for method invocation.
```
`//`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+                                                           dynamicPath.variables());
+            if (!missingVariables.isEmpty()) {
+                String msg = String.format("%s does not contains following variables: %s", mappedTo,
+                                           missingVariables.stream().collect(Collectors.joining(", ")));
```
No need to use a joiner?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,165 @@
+ * @see DynamicHttpService
+ * @see Path
+ */
+public final class DynamicPath {
```
Maybe this class could be package-local? What do you think?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/MappedDynamicFunction.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * {@link DynamicHttpFunction} with args.
+ */
+class MappedDynamicFunction extends AbstractHttpService {
```
`final`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Returns the array of {@link Path} annotated methods.
+     */
+    public static Method[] requestMappingMethods(Object object) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Returns the array of parameter types.
+     */
+    public static Class<?>[] parameterTypes(Method method) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Returns the array of path variable values of parameters.
+     */
+    public static String[] pathVariables(Method method) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+     * @see Trace
+     * @see Connect
+     */
+    public static EnumSet<HttpMethod> httpMethods(Method method) {
```
`public` could be removed.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+        int parameterCount = method.getParameterCount();
+        return Arrays.stream(method.getParameters())
+                     .map(p -> p.getType())
+                     .toArray(size -> new Class<?>[parameterCount]);
```
Does `Class[]::new` work here?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+        }
+
+        if (ret.isEmpty()) {
+            ret = EnumSet.allOf(HttpMethod.class);
```
Perhaps we should skip the methods that are not annotated with `@<method>` annotations rather than accepting all HTTP methods?

dongjinleekr replied:
No problem. Applying this, I added some validation routine for checking @Path annotated method without any HTTP Method annitations, e.g., @Get, @Post, etc in DynamicHttpServiceBuilder#add(Object).

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+            String[] pathVariables = Methods.pathVariables(method);
+            assert parameterCount == parameterTypes.length;
+            assert parameterCount == pathVariables.length;
+            Set<String> missingVariables = Sets.difference(Sets.newHashSet(pathVariables),
```
Which is better, `Sets.newHashSet(pathVariables)` or `new HashSet<>(pathVariables)`?

dongjinleekr replied:
Hmm, I see. I fix all the other Guava container methods. It seems like you prefer specifying the container type explicitly. (or, using container constructor explicitly.)

trustin replied:
The thing is that we prefer less indirect form when they are practically identical with same amount of code.

kojilin replied:
Even guava suggest using new diamond syntax.

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,191 @@
+            String mappedTo = mapping.value();
+            DynamicPath dynamicPath = DynamicPath.of(mappedTo);
+            int parameterCount = method.getParameterCount();
+            Class<?>[] parameterTypes = Methods.parameterTypes(method);
```
Think we can merge `parameterTypes()` and `pathVariables()` into one.

kojilin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+        return Arrays.stream(method.getParameters())
+                     .map(p -> p.getAnnotation(PathParam.class))
+                     .map(a -> a.value())
+                     .toArray(size -> new String[parameterCount]);
```
`.toArray(String[]::new)` an u don't need `int parameterCount = method.getParameterCount();` I think.

kojilin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+    public static Method[] requestMappingMethods(Object object) {
+        List<Method> methods = Arrays.stream(object.getClass().getMethods())
+                                     .filter(m -> null != m.getAnnotation(Path.class))
+                                     .collect(Collectors.toList());
```
just toArray here.

kojilin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,126 @@
+     * </ul>
+     */
+    private static final Map<Class<?>, HttpMethod> HTTP_METHOD_MAP =
+            ImmutableMap.<Class<?>, HttpMethod>builder()
```
` ImmutableMap.builder()` is enough?

dongjinleekr commented:
Here is the update with full features: I added support for Content conversion and asynchronous method call, by making DynamicHttpFunction#server return anything. Instead, I added some internal utility methods which compose given user-defined DynamicHttpFunction and ResponseConverters into directly invokable DynamicHttpFunction object - see `DynamicHttpFunctions#of` methods. Now the overall API looks like [ResponseTransformer of SparkJava](http://sparkjava.com/documentation.html#response-transformer).

Asynchronous method call also works similarly. For this feature, please see `DynamicHttpFunctions#async` methods.

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/ParameterEntry.java:
```diff
@@ -0,0 +1,24 @@
+        this.name = name;
+    }
+
+    public Class<?> getType() {
```
Can remove private?

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java:
```diff
@@ -0,0 +1,75 @@
+        for (int i = 0; i < parameterEntries.length; ++i) {
+            ParameterEntry entry = parameterEntries[i];
+            String variable = entry.getName();
+            assert args.containsKey(variable);
```
I think it is better to check result of `args.get(variable)` instead of `containsKey`.
```java
String value = args.get(variable); 
assert value != null;
```

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,92 @@
+     *
+     * @see MappedDynamicFunction
+     */
+    MappedDynamicFunction bind(HttpMethod method, String mappedPath) {
```
Add `@Nullable`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    private static boolean deserializable(Class<?> clazz) {
```
`canDeserialize`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunction.java:
```diff
@@ -0,0 +1,17 @@
+@FunctionalInterface
+public interface DynamicHttpFunction {
+
+    Object serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) throws Exception;
```
Javadoc?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,92 @@
+    private final DynamicHttpFunction function;
+
+    /**
+     * Creates a instance.
```
a new instance

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Set of HTTP Method, e.g., GET, POST, PUT, ...
+     */
+    private final EnumSet<HttpMethod> methods;
```
`Set`

trustin replied:
No, I meant 'please change the type of `methods` from `EnumSet<HttpMethod>` to `Set<HttpMethod>`. I don't see a reason to declare it as an `EnumSet`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Creates a instance.
+     */
+    DynamicHttpFunctionEntry(EnumSet<HttpMethod> methods, DynamicPath path, DynamicHttpFunction function) {
```
- Could you make a defensive copy of `methods`? e.g. `EnumSet.copyOf(methods)`
- `EnumSet` -> `Set`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,92 @@
+     * Creates a instance.
+     */
+    DynamicHttpFunctionEntry(EnumSet<HttpMethod> methods, DynamicPath path, DynamicHttpFunction function) {
+        this.methods = methods;
```
missing `requireNonNull`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,133 @@
+        }
+
+        // search for the converter mapped to one of its interface.
+        for (Class<?> itfc : type.getInterfaces()) {
```
You also need to look for the interfaces of superclasses. See `ThriftServiceMetadata.getAllInterfaces()`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,133 @@
+/**
+ * Provides various utility functions for internal use related with {@link DynamicHttpFunction}.
+ */
+final class DynamicHttpFunctions {
```
Is this class meant to be `public`? I see some public methods in this class.

dongjinleekr replied:
> Is this class meant to be `public`?

No. I will remove all `public` modifiers from the methods. It is intended to be a pure internal class.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,133 @@
+     * Creates a new {@link DynamicHttpFunction} object which runs given {@code function} asynchronously
+     * with {@code executor}.
+     */
+    public static DynamicHttpFunction async(DynamicHttpFunction function, Executor executor) {
```
`ofAsync`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,232 @@
+ */
+public final class DynamicHttpServiceBuilder {
+
+    private Executor executor = ForkJoinPool.commonPool();
```
We already have `ServiceRequestContext.blockingTaskExecutor()`. No need to have an `executor` property.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,232 @@
+    /**
+     * Adds a mapping from {@link Class} to {@link ResponseConverter}.
+     */
+    public DynamicHttpServiceBuilder convert(Class<?> type, ResponseConverter converter) {
```
- `converter`?
- How about adding a method like `ResponseConverter.supportedTypes()` so that a user doesn't need to specify `type` at all?

dongjinleekr replied:
- The reason I used `convert` instead of `convert` was that imperative form would be better: "convert `type` using given `converter` if no explicit converter is given." Would it better to change into `converter`?
- Since `ResponseConverter` is a functional interface, it is impossible to add a method, which requires a internal variable. It is why I made to specify the class type explicitly.

trustin replied:
Yes, `converter` or `addConverter` sound OK to me. A builder is often meant to keep add something until `build()`, so I think it's fine to omit `set` or `add` prefix.

trustin replied:
> Since `ResponseConverter` is a functional interface, it is impossible to add a method.

Could you explain why it has to remain to be a functional interface?

dongjinleekr replied:
> Yes, `converter` or `addConverter` sound OK to me. A builder is often meant to keep add something until `build()`, so I think it's fine to omit `set` or `add` prefix.

Understood. I changed the method name into `addConverter`, which corresponds with `addMapping(s)` in this update.

> Could you explain why it has to remain to be a functional interface?

The reason I decided to make it a functional interface is that It would be better if it can be defined using lambda. That's all.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,232 @@
+
+            DynamicHttpFunctionEntry entry = new DynamicHttpFunctionEntry(methods, dynamicPath,
+                                                                          DynamicHttpFunctions
+                                                                                  .of(function, converters));
```
Could look prettier this way:

    DynamicHttpFunctionEntry entry = new DynamicHttpFunctionEntry(
            methods, dynamicPath, DynamicHttpFunctions.of(function, converters));

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpServiceBuilder.java:
```diff
@@ -0,0 +1,232 @@
+     * Adds a new dynamic-mapped method, invoked with given {@link HttpMethod} and (dynamically-bound) path,
+     * runs asynchronously with the default {@link Executor}.
+     */
+    public DynamicHttpServiceBuilder async(HttpMethod method, String path, DynamicHttpFunction function) {
```
addAsyncMapping?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,164 @@
+ * @see DynamicHttpService
+ * @see Path
+ */
+final class DynamicPath {
```
Is this class meant to be public? If not, please remove the `public` modifiers from the methods in this class.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/ResponseConverter.java:
```diff
@@ -0,0 +1,28 @@
+@FunctionalInterface
+public interface ResponseConverter {
+
+    HttpResponse handle(Object object) throws Exception;
```
- handle -> convert?
- object -> resObj?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/ResponseConverter.java:
```diff
@@ -0,0 +1,28 @@
+import com.linecorp.armeria.common.http.HttpResponse;
+
+/**
+ * Converts given {@code T} type object into {@link HttpResponse}.
```
There's no type parameter in this class, but you could add one?

dongjinleekr replied:
Oh my god, it is a residue I left when I was trying to make `ResponseConverter` a generic interface. After some meaningless efforts, I decided to make it non-generic. I will fix the Javadoc comments.

dongjinleekr commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,68 @@
+            return (T) function.apply(str);
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(
+                    String.format("Can't deserialize %s to type %s", str, clazz.getName()));
```
After reviewing the usages from the project, I concluded that `Class#getSimpleName()` is more common in Exception messages than `Class#getName()`. But It does not seem there is some rule - some exception messages (about three only) are using `Class#getName()`.

I think using `Class#getSimpleName()` would be better for the exception messages. If you agree, I will fix the exception messages that is using `Class#getName()`.

trustin replied:
+1 for `getSimpleName()`

trustin replied:
Still not using `getSimpleName()`?

dongjinleekr commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,164 @@
+            return null;
+        }
+
+        ImmutableMap.Builder<String, String> builder = ImmutableMap.<String, String>builder();
```
I considered using the stream for the first time but changed my mind not to using it because `DynamicPath#bind` is called so frequently.

dongjinleekr commented:
Here is the update. The heart of this update is `MappedDynamicFunction#serve`, which runs `DynamicHttpFunction` using `ServiceRequestContext.blockingTaskExecutor()`. However, since I have no experience on RxJava Streams, I cannot certain that there is a better way to deliver HttpResponse from working thread to IO thread. Is this OK?

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/Types.java:
```diff
@@ -0,0 +1,49 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
```
Could we move this class to `com.linecorp.armeria.internal`? I don't want Armeria to provide generic utilities that are not related with how users use Armeria.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Get.java:
```diff
@@ -0,0 +1,32 @@
+import com.linecorp.armeria.common.http.HttpMethod;
+
+/**
+ * Class for mapping {@link HttpMethod#GET} HTTP Method onto specific method.
```
Global comment:

- Class -> Annotation
- Remove 'HTTP Method' because the `{@link HttpMethod#...}` already tells a user that it's about an HTTP method.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/ResponseConverter.java:
```diff
@@ -0,0 +1,28 @@
+@FunctionalInterface
+public interface ResponseConverter {
+
+    HttpResponse convert(Object resObj) throws Exception;
```
Missing Javadoc.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,117 @@
+                                  Map<Class<?>, ResponseConverter> converters) {
+        return (ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) -> {
+            Object ret = function.serve(ctx, req, args);
+            if (CompletionStage.class.isAssignableFrom(ret.getClass())) {
```
No need to use `isAssignableFrom()`:

```java
if (ret instanceof CompletionStage) {
    ...
}
```

Also, please check if ret is an `HttpResponse` already so that conversion is bypassed simply.

dongjinleekr replied:
The case when `ret` is an instance of `HttpResponse` is handled by `convert` method, which uses `isAssignableFrom()` also. So, I decided to fix both methods to use `instanceof` instead. Is this okay?

trustin replied:
OK

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    static DynamicHttpFunction of(DynamicHttpFunction function, ResponseConverter converter) {
+        return (ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) -> {
+            Object ret = function.serve(ctx, req, args);
```
You should handle the case where `ret` is `instanceof CompletionStage` or `instanceof HttpResponse` here as well.

trustin commented:
> I cannot certain that there is a better way to deliver HttpResponse from working thread to IO thread. Is this OK?

It is still blocking, but let's forget about how to make this truly asynchronous at the moment.

dongjinleekr commented:
Here is the update:

1. Move `Type` class into `com.linecorp.armeria.internal` package.
2. Fix loggings on `Deserializers`: `Class#getName` -> `Class#getSimpleName`.
3. Fix javadocs: `Delete`, `Get`, `Head`, `Options`, `Patch`, `Post`, `Put`, `Trace` and `ResponseConverter#convert`.
4. Update `DynamicMethods`:
    - Add `convert(Object, ResponseConverter)` to handle the case of explicit converter is given.
    - Add asynchronous method call support to DynamicHttpFunctions#of(DynamicHttpFunction, ResponseConverter). This method now also handles the returned object of `HttpResponse` type, using `convert` method above.
    - Change `isAssignableFrom()` into `instanceof`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,143 @@
+            try {
+                return converter.convert(object);
+            } catch (Exception e) {
+                throw new IllegalStateException("Exception occurred during ResponseConverter#convert");
```
Should specify the cause when creating the exception.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    public static <T> T deserialize(String str, Class<T> clazz) {
+        if (!canDeserialize(clazz)) {
+            throw new IllegalArgumentException("Class " + clazz.getSimpleName() + "can't be deserialized.");
```
`"can't` -> `" can't`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,95 @@
+     * Creates a new instance.
+     */
+    DynamicHttpFunctionEntry(EnumSet<HttpMethod> methods, DynamicPath path, DynamicHttpFunction function) {
+        this.methods = EnumSet.copyOf(requireNonNull(methods));
```
`requireNonNull(methods, "methods")`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    boolean overlaps(DynamicHttpFunctionEntry entry) {
+        return !Sets.intersection(methods, entry.methods).isEmpty() && path.skeleton().equals(
+                entry.path.skeleton());
```
Could be prettier:

    return !Sets.intersection(methods, entry.methods).isEmpty() &&
           path.skeleton().equals(entry.path.skeleton());

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Returns the set of parameter names, which should be provided to invoke this function.
+     */
+    public Set<String> parameterNames() {
```
`public` is redundant.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,143 @@
+
+        // No appropriate converter found: raise runtime exception.
+        throw new IllegalArgumentException(
+                "No available converter exists for " + type.getSimpleName() + " type");
```
`"converter not available for " + type.getSimpleName()"`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctions.java:
```diff
@@ -0,0 +1,143 @@
+            try {
+                return converter.convert(object);
+            } catch (Exception e) {
+                throw new IllegalStateException("Exception occurred during ResponseConverter#convert");
```
Cause should be specified

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/MappedDynamicFunction.java:
```diff
@@ -0,0 +1,60 @@
+        ctx.blockingTaskExecutor().execute(() -> {
+            try {
+                Object ret = function.serve(ctx, req, args);
+                if (!HttpResponse.class.isAssignableFrom(ret.getClass())) {
```
`instanceof`

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,68 @@
+            return (T) function.apply(str);
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(
+                    "Can't deserialize " + str + " to type " + clazz.getSimpleName());
```
- No need to catch `Throwable`. `Exception` should be enough.
- Missing cause

dongjinleekr replied:
Fixed.

trustin commented:
@imasahiro Care to take a final review?

dongjinleekr commented:
![sphinx](https://cloud.githubusercontent.com/assets/2375128/21477657/20595a1e-cb87-11e6-8c14-f557ee100cce.png)

Hmm... Why the Travis-ci job was failed? It succeeds with my local dev machine.

trustin commented:
No idea. It seems unrelated with your work. Please ignore it for now.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpService.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * An {@link HttpService} that serves dynamic contents.
+ */
+public final class DynamicHttpService extends AbstractHttpService {
```
How about:

- making this class non-final,
- providing a protected default constructor which populates the entries using `this` as the strategy

..? A user could build a dynamic service like this:

```java
public class MyDynamicHttpService extends DynamicHttpService {
    @Get
    @Path(....)
    public Object someMethod(...) { ... }
}
```

dongjinleekr replied:
Let me summarize. In that case, the user can build a dynamic service in 3 ways like following:

1. Create `DynamicHttpFunction function`, `ResponseConverter` instances and use `DynamicHttpServiceBuilder#addMapping`.
2. Create arbitrary-typed a strategy class with annotations and use `DynamicHttpServiceBuilder#addMappings`.
3. Create DynamicHttpService directly by extending DynamicHttpService and use annotations.

Is this what you mean?

trustin replied:
Yes. You implemented (1) and (2) already, right? I'm suggesting to add (3).

dongjinleekr replied:
I understood. But with the 3rd approach, there is no way to specify `ResponseConverter` to the annotated methods yet. How about add a new annotation to specify `ResponseConverter`? In that case, your suggestion would be updated like following:

```
public class MyDynamicHttpService extends DynamicHttpService {
    @Get
    @Path(....)
    @Response(NumberConverter.class)
    public long someMethod(...) { ... }
}
```

dongjinleekr replied:
Add to this, I think approach 2 and 3 are too similar each other. How about replace approach 2 with approach 3? It would be much simpler and better. With this way, we can have 2 ways of building `DynamicHttpService`:

1. Create `DynamicHttpService` indirectly using `DynamicHttpServiceBuilder`.
2. Create `DynamicHttpService` directly by extending & annotating the class. (without `DynamicHttpServiceBuilder`)

trustin replied:
- +1 for allowing to specify a `ResponseConverter` in an annotation. 
- I think (2) is different from (3) because (2) allows a user specify multiple strategy objects.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,107 @@
+     *   <li> {@link Put} -> {@link HttpMethod#PUT}
+     *   <li> {@link Patch} -> {@link HttpMethod#PATCH}
+     *   <li> {@link Delete} -> {@link HttpMethod#DELETE}
+     *   <li> {@link Trace} -> {@link HttpMethod#TRACE}
```
`<li> {` -> `<li>{`

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,167 @@
+    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/]+)+");
+
+    /**
+     * Create a {@link DynamicPath} instance from given str.
```
nit. How about rename `str` to `path`? and update javadoc from `from given str` to `from given {@code path}`?

dongjinleekr replied:
No problem.

dongjinleekr replied:
I also fixed the constructor of `DynamicPath`, along with the static function you mentioned.

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,167 @@
+            return null;
+        }
+
+        ImmutableMap.Builder<String, String> builder = ImmutableMap.<String, String>builder();
```
I see. Then could you remove type arguments from `builder()`? I think `ImmutableMap.builder()` is enough.

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,107 @@
+        int parameterCount = method.getParameterCount();
+        return Arrays.stream(method.getParameters())
+                     .map(p -> new ParameterEntry(p.getType(), p.getAnnotation(PathParam.class).value()))
+                     .toArray(size -> new ParameterEntry[parameterCount]);
```
Can use method reference `ParameterEntry[]::new`

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,107 @@
+    static Method[] requestMappingMethods(Object object) {
+        List<Method> methods = Arrays.stream(object.getClass().getMethods())
+                                     .filter(m -> m.getAnnotation(Path.class) != null)
+                                     .collect(Collectors.toList());
```
How about using `toArray(Method[]::new)`?

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,68 @@
+    private static final Map<Class<?>, Function<String, Object>> FUNCTIONS;
+
+    static {
+        FUNCTIONS = new HashMap<>();
```
Is it better to make `FUNCTIONS` immutable?

dongjinleekr commented:
Here is the update, with the squashed commits:

1. Fix Javadoc: `Path`, `PathParam`.
2. Add new annotations: `Convert`, `Converter` and `Converters`.
3. Make `Deserializers#FUNCTIONS` immutable.
4. Make `DynamicHttpService` non-final, making the 3rd way of defining Dynamic HTTP service.

Add to this, I re-organized test methods in `HttpServiceTest#testDynamicHttpServices`. As you know, there are three ways of configuring Dynamic HTTP service. I organized test cases of 9, like following:

- case 1, 2, 3 for using `DynamicHttpServiceBuilder#addMapping` and `DynamicHttpFunction` functional interface.
- case 4, 5, 6 for using `DynamicHttpServiceBuilder#addMappings` and strategy object.
- case 7, 8, 9 for defining service object by extending `DynamicHttpService`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Converters.java:
```diff
@@ -0,0 +1,34 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Converters {
```
Is this annotation necessary? I think a user can just repeat `@Converter` without this:

```java
@Converter(...)
@Converter(...)
public class ... { ... }
```

.. and you can get the annotations using `Class.getAnnotationsByType()`.

dongjinleekr replied:
After reviewing the documentations, I found that we need container annotation to repeat the same annotations - without it, a compilation occurs when a user tries to repeat the annotation. It is why I decided that `Converters` annotation is necessary.

- [Repeating Annotations (Oracle official Java guide)](http://docs.oracle.com/javase/tutorial/java/annotations/repeating.html)
- [Java 8 Repeating Annotation Explained in 5 minutes](https://blog.idrsolutions.com/2015/03/java-8-repeating-annotation-explained-in-5-minutes/)

However, after thinking carefully about your comment, I agree that `Converters` don't need to go public. Then, how about make this annotation package-private?

trustin replied:
Ah, I see. Thanks for the links. 

> However, after thinking carefully about your comment, I agree that Converters don't need to go public. Then, how about make this annotation package-private?

+1

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Convert.java:
```diff
@@ -0,0 +1,34 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface Convert {
```
I'm not sure I like that we have `@Convert` and `@Converter`. How about:

- removing `@Convert`
- setting the default value of `Converter.target` to something like `Object.class`
- rejecting the `@Converter` annotation that annotates a class and has 'Object.class' target ..?

dongjinleekr replied:
If we remove `@Convert`, there is no way to specify `ResponseConverter` to the methods. Is it okay? The relationship between `@Convert` and `@Converter` is similar to `DynamicHttpServiceBuilder#addMapping(Iterable<HttpMethod>, String, DynamicHttpFunction, ResponseConverter)` and `DynamicHttpServiceBuilder#addConverter`.

dongjinleekr replied:
Or, would it better to remove `@Convert` and make `@Converter` to annotate methods only? If it is your intention, we can't configure class-wide `ResponseConverter`. Is it okay?

trustin replied:
You can specify multiple targets for an annotation:

```java
@Target({ElementType.TYPE, ElementType.METHOD})
```

dongjinleekr replied:
Understood. It seems like I have to update the annotation-processing routine. I will handle this issue along with the above problem, `Converters`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    /**
+     * Returns the {@link ResponseConverter} instance from {@Convert} annotation of the given {@code method}.
```
Global comment:
- `{@Convert}` -> `{@link Convert}`
- `{@Converter}` -> `{@link Converter}`

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,167 @@
+            String variable = variable(token);
+            if (variable != null) {
+                if (!placeholders.contains(variable)) {
+                    /**
```
Replace javadoc style multi-line comment (`/** ... */`)to `//`.

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,167 @@
+                    patternJoiner.add(String.format("(?<%s>[^/]+)", variable));
+                } else {
+                    /**
+                     * If given token is a re-occurred variable mark, replace it to backreference expression
```
ditto

imasahiro commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java:
```diff
@@ -0,0 +1,167 @@
+                }
+                skeletonJoiner.add("{}");
+            } else {
+                /**
```
`/** ... */` -> `// If given token is ...`

dongjinleekr commented:
Here is the update:

1. The functionality of `@Convert` annotation was merged into `@Converter`. (proposed by @trustin) In other words, `@Convert` was removed and the target type of `@Converter` was expanded.
2. All Javadoc-style multiline comments were replaced with `//`. (proposed by @imasahiro)

Followings are some remarkables:

1. For the annotation's default value should be a constant, it can't be null. So, I decided for the default value of `Converter.target` to be `Object.class`.
2. The usage of Converter annotation is a little bit different for Methods and Classes. When it is marked on a method, it can't be repeated and should not specify the target class. But if it is marked on a class, it can be repeated and must specify the target class, except `Object.class`.
3. I tried to make `@Converters` to package-private, but it was impossible. So I left it unchanged.

Add to this; the main updates reside on `Methods#converter` and `Methods#converters`.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Deserializers.java:
```diff
@@ -0,0 +1,67 @@
+                    .put(Long.TYPE, Long::parseLong)
+                    .put(Float.TYPE, Float::parseFloat)
+                    .put(Double.TYPE, Double::parseDouble)
+                    .put(String.class, s -> s)
```
Could this be `Function.identity()`? Not completely sure whether the compiler will complain or not

dongjinleekr replied:
I already tried it, but it results in a compilation error.

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/Methods.java:
```diff
@@ -0,0 +1,217 @@
+            Class<?> target = converter.target();
+            if (target == Object.class) {
+                throw new IllegalArgumentException(
+                        "@Converter annotation can't be marked on a class without a target specified.");
```
- can't be marked on a class without a target specified -> must have a target type specified

dongjinleekr commented:
Fixed Javadocs & Exception Messages:

1. `Converter#value`: `ResponseConverter class.` -> `The type of the {@link ResponseConverter} that will convert the object whose type is {@link #target()}`
2. `Converters#value`: `Array of {@link Converter}.` -> `An array of {@link Converter}s.`
3. `Methods#converters`: `@Converter annotation can't be marked on a class without a target specified.` -> `@Converter annotation must have a target type specified.`

imasahiro commented:
LGTM. BTW, do we need to update copyright header from 2016 to 2017?
`Copyright 2016 LINE Corporation`

trustin commented:
Thank you very much for your patience and hard work, @dongjinleekr !

trustin commented:
@imasahiro Ugh, I forgot to update the years.. Maybe that's fine.. :stuck_out_tongue_closed_eyes:
